base/applications/rapps/misc.cpp

```diff
@@ -368,6 +347,64 @@ UnixTimeToFileTime(DWORD dwUnixTime, LPFILETIME pFileTime)
     pFileTime->dwHighDateTime = ll >> 32;
 }
 
+HRESULT
+RegKeyHasValues(HKEY hKey, LPCWSTR Path, REGSAM wowsam)
+{
+    CRegKey key;
+    LONG err = key.Open(hKey, Path, KEY_QUERY_VALUE | wowsam);
+    if (err == ERROR_SUCCESS)
+    {
+        WCHAR name[1];
+        DWORD cchname = _countof(name), cbsize = 0;
+        err = RegEnumValueW(key, 0, name, &cchname, NULL, NULL, NULL, &cbsize);
+        if (err == ERROR_NO_MORE_ITEMS)
+            return S_FALSE;
+        if (err == ERROR_MORE_DATA)
+            err = ERROR_SUCCESS;
+    }
+    return HRESULT_FROM_WIN32(err);
+}
+
+LPCWSTR
+GetRegString(CRegKey &Key, LPCWSTR Name, CStringW &Value)
+{
+    for (;;)
+    {
+        ULONG cb = 0, cch;
+        ULONG err = Key.QueryValue(Name, NULL, NULL, &cb);
+        if (err)
+            break;
+        cch = cb / sizeof(WCHAR);
+        LPWSTR p = Value.GetBuffer(cch + 1);
+        p[cch] = UNICODE_NULL;
+        err = Key.QueryValue(Name, NULL, (BYTE*)p, &cb);
+        if (err == ERROR_MORE_DATA)
```

Comment:
Is this in order to trying to workaround concurrent access to this registry value, where its data could grow between the QueryValue call from above, and the allocation and the QueryValue call from here?

Reply:
Yes. (technically all code that reads from the registry should have a loop like this)

For Rapps, it's conceivable that someone is refreshing the UI while waiting for an installer that is running in the background. Ideally all code that reads from the uninstall keys should use a loop.